instrumentation/lettuce/lettuce-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/lettuce/v4_0/LettuceAsyncClientTest.java

```diff
@@ -0,0 +1,568 @@
+    testConnectionClient.setOptions(CLIENT_OPTIONS);
+
+    StatefulRedisConnection<String, String> connection =
+        testConnectionClient.connect(
+            new Utf8StringCodec(), new RedisURI(host, port, 3, TimeUnit.SECONDS));
+
+    assertThat(connection).isNotNull();
+
+    testing.waitAndAssertTraces(
+        trace ->
+            trace.hasSpansSatisfyingExactly(
+                span ->
+                    span.hasName("CONNECT")
+                        .hasKind(SpanKind.CLIENT)
+                        .hasAttributesSatisfyingExactly(
+                            equalTo(SemanticAttributes.NET_PEER_NAME, host),
+                            equalTo(SemanticAttributes.NET_PEER_PORT, port),
+                            equalTo(SemanticAttributes.DB_SYSTEM, "redis"))));
+    connection.close();
+  }
+
+  @Test
+  void testExceptionInsideTheConnectionFuture() {
+    RedisClient testConnectionClient = RedisClient.create(dbUriNonExistent);
+    testConnectionClient.setOptions(CLIENT_OPTIONS);
+
+    Exception exception =
+        catchException(
+            () -> {
+              StatefulRedisConnection<String, String> connection =
+                  testConnectionClient.connect(
+                      new Utf8StringCodec(),
+                      new RedisURI(host, incorrectPort, 3, TimeUnit.SECONDS));
+
+              assertThat(connection).isNull();
+            });
+
+    assertThat(exception).isInstanceOf(RedisConnectionException.class);
+
+    testing.waitAndAssertTraces(
+        trace ->
+            trace.hasSpansSatisfyingExactly(
+                span ->
+                    span.hasName("CONNECT")
+                        .hasKind(SpanKind.CLIENT)
+                        .hasStatus(StatusData.error())
+                        .hasException(exception)
+                        .hasAttributesSatisfyingExactly(
+                            equalTo(SemanticAttributes.NET_PEER_NAME, host),
+                            equalTo(SemanticAttributes.NET_PEER_PORT, incorrectPort),
+                            equalTo(SemanticAttributes.DB_SYSTEM, "redis"))));
+  }
+
+  @Test
+  void testSetCommandUsingFutureGetWithTimeout()
+      throws ExecutionException, InterruptedException, TimeoutException {
+    RedisFuture<String> redisFuture = asyncCommands.set("TESTSETKEY", "TESTSETVAL");
+    String res = redisFuture.get(3, TimeUnit.SECONDS);
+
+    assertThat(res).isEqualTo("OK");
+
+    testing.waitAndAssertTraces(
+        trace ->
+            trace.hasSpansSatisfyingExactly(
+                span ->
+                    span.hasName("SET")
+                        .hasKind(SpanKind.CLIENT)
+                        .hasAttributesSatisfyingExactly(
+                            equalTo(SemanticAttributes.DB_SYSTEM, "redis"),
+                            equalTo(SemanticAttributes.DB_OPERATION, "SET"))));
+  }
+
+  @Test
+  void testCommandChainedWithThenAccept() throws Throwable {
+    AsyncConditions conditions = new AsyncConditions();
```

Comment:
Instead of that you could use either an `AtomicReference` or a `CompletableFuture` that you would set from the `redisFuture.thenAccept()` callback; and then assert on that ref/future using `Awaitility.await()`

Reply:
thank you for this suggestion, I believe I have implemented the proposed change